--- a/net/minecraft/server/commands/WeatherCommand.java
+++ b/net/minecraft/server/commands/WeatherCommand.java
@@ -48,20 +_,26 @@
     }
 
     private static int setClear(CommandSourceStack source, int time) {
+        io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTask(() -> { // Folia - region threading
         source.getLevel().setWeatherParameters(getDuration(source, time, ServerLevel.RAIN_DELAY), 0, false, false); // CraftBukkit - SPIGOT-7680: per-world
         source.sendSuccess(() -> Component.translatable("commands.weather.set.clear"), true);
+        }); // Folia - region threading
         return time;
     }
 
     private static int setRain(CommandSourceStack source, int time) {
+        io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTask(() -> { // Folia - region threading
         source.getLevel().setWeatherParameters(0, getDuration(source, time, ServerLevel.RAIN_DURATION), true, false); // CraftBukkit - SPIGOT-7680: per-world
         source.sendSuccess(() -> Component.translatable("commands.weather.set.rain"), true);
+        }); // Folia - region threading
         return time;
     }
 
     private static int setThunder(CommandSourceStack source, int time) {
+        io.papermc.paper.threadedregions.RegionizedServer.getInstance().addTask(() -> { // Folia - region threading
         source.getLevel().setWeatherParameters(0, getDuration(source, time, ServerLevel.THUNDER_DURATION), true, true); // CraftBukkit - SPIGOT-7680: per-world
         source.sendSuccess(() -> Component.translatable("commands.weather.set.thunder"), true);
+        }); // Folia - region threading
         return time;
     }
 }
