--- a/src/main/java/io/papermc/paper/entity/activation/ActivationType.java
+++ b/src/main/java/io/papermc/paper/entity/activation/ActivationType.java
@@ -19,7 +_,7 @@
     RAIDER,
     MISC;
 
-    AABB boundingBox = new AABB(0, 0, 0, 0, 0, 0);
+    //AABB boundingBox = new AABB(0, 0, 0, 0, 0, 0); // Folia - threaded regions - replaced by local variable
 
     /**
      * Returns the activation type for the given entity.
