--- a/net/minecraft/world/damagesource/DamageSource.java
+++ b/net/minecraft/world/damagesource/DamageSource.java
@@ -163,12 +_,12 @@
         if (this.causingEntity == null && this.directEntity == null) {
             LivingEntity killCredit = livingEntity.getKillCredit();
             String string1 = string + ".player";
-            return killCredit != null
+            return killCredit != null && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(killCredit)
                 ? Component.translatable(string1, livingEntity.getDisplayName(), killCredit.getDisplayName())
                 : Component.translatable(string, livingEntity.getDisplayName());
         } else {
             Component component = this.causingEntity == null ? this.directEntity.getDisplayName() : this.causingEntity.getDisplayName();
-            ItemStack itemStack = this.causingEntity instanceof LivingEntity livingEntity1 ? livingEntity1.getMainHandItem() : ItemStack.EMPTY;
+            ItemStack itemStack = this.causingEntity instanceof LivingEntity livingEntity1 && ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(livingEntity1) ? livingEntity1.getMainHandItem() : ItemStack.EMPTY; // Folia - region threading
             return !itemStack.isEmpty() && itemStack.has(DataComponents.CUSTOM_NAME)
                 ? Component.translatable(string + ".item", livingEntity.getDisplayName(), component, itemStack.getDisplayName())
                 : Component.translatable(string, livingEntity.getDisplayName(), component);
