--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMushroomCow.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMushroomCow.java
@@ -19,6 +_,13 @@
         super(server, entity);
     }
 
+    // Folia start - region threading
+    @Override
+    public net.minecraft.world.entity.animal.MushroomCow getHandleRaw() {
+        return (net.minecraft.world.entity.animal.MushroomCow)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public boolean hasEffectsForNextStew() {
         SuspiciousStewEffects stewEffects = this.getHandle().stewEffects;
@@ -94,6 +_,7 @@
 
     @Override
     public net.minecraft.world.entity.animal.MushroomCow getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (net.minecraft.world.entity.animal.MushroomCow) this.entity;
     }
 
