--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -306,8 +_,16 @@
         this.mode = mode;
     }
 
+    // Folia start - region threading
+    @Override
+    public Player getHandleRaw() {
+        return (Player)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public Player getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (Player) this.entity;
     }
 
