--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftCreeper.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftCreeper.java
@@ -87,6 +_,13 @@
         this.getHandle().ignite();
     }
 
+    // Folia start - region threading
+    @Override
+    public net.minecraft.world.entity.monster.Creeper getHandleRaw() {
+        return (net.minecraft.world.entity.monster.Creeper)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public Entity getIgniter() {
         return (this.getHandle().entityIgniter != null) ? this.getHandle().entityIgniter.getBukkitEntity() : null;
@@ -94,6 +_,7 @@
 
     @Override
     public net.minecraft.world.entity.monster.Creeper getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (net.minecraft.world.entity.monster.Creeper) this.entity;
     }
 
