--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftTNTPrimed.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftTNTPrimed.java
@@ -42,8 +_,16 @@
         this.getHandle().setFuse(fuseTicks);
     }
 
+    // Folia start - region threading
+    @Override
+    public PrimedTnt getHandleRaw() {
+        return (PrimedTnt)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public PrimedTnt getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (PrimedTnt) this.entity;
     }
 
