--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFireball.java
@@ -83,8 +_,16 @@
     }
     // Paper end - Expose power on fireball projectiles
 
+    // Folia start - region threading
+    @Override
+    public AbstractHurtingProjectile getHandleRaw() {
+        return (AbstractHurtingProjectile)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public AbstractHurtingProjectile getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (AbstractHurtingProjectile) this.entity;
     }
 
