--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftIllager.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftIllager.java
@@ -10,8 +_,16 @@
         super(server, entity);
     }
 
+    // Folia start - region threading
+    @Override
+    public AbstractIllager getHandleRaw() {
+        return (AbstractIllager)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public AbstractIllager getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (AbstractIllager) super.getHandle();
     }
 
