--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftThrowableProjectile.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftThrowableProjectile.java
@@ -26,8 +_,16 @@
         this.getHandle().setItem(CraftItemStack.asNMSCopy(item));
     }
 
+    // Folia start - region threading
+    @Override
+    public ThrowableItemProjectile getHandleRaw() {
+        return (ThrowableItemProjectile)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public ThrowableItemProjectile getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (ThrowableItemProjectile) this.entity;
     }
 }
