--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecart.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecart.java
@@ -77,8 +_,16 @@
     }
     // Paper end
 
+    // Folia start - region threading
+    @Override
+    public AbstractMinecart getHandleRaw() {
+        return (AbstractMinecart)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public AbstractMinecart getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (AbstractMinecart) this.entity;
     }
 
