--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftWarden.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftWarden.java
@@ -15,8 +_,16 @@
         super(server, entity);
     }
 
+    // Folia start - region threading
+    @Override
+    public Warden getHandleRaw() {
+        return (Warden)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public Warden getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (Warden) this.entity;
     }
 
