--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftTrident.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftTrident.java
@@ -12,8 +_,16 @@
         super(server, entity);
     }
 
+    // Folia start - region threading
+    @Override
+    public ThrownTrident getHandleRaw() {
+        return (ThrownTrident)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public ThrownTrident getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (ThrownTrident) super.getHandle();
     }
 
