--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftBat.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftBat.java
@@ -8,8 +_,16 @@
         super(server, entity);
     }
 
+    // Folia start - region threading
+    @Override
+    public net.minecraft.world.entity.ambient.Bat getHandleRaw() {
+        return (net.minecraft.world.entity.ambient.Bat)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public net.minecraft.world.entity.ambient.Bat getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (net.minecraft.world.entity.ambient.Bat) this.entity;
     }
 
