--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPiglin.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPiglin.java
@@ -75,8 +_,16 @@
         return new CraftInventory(this.getHandle().inventory);
     }
 
+    // Folia start - region threading
+    @Override
+    public net.minecraft.world.entity.monster.piglin.Piglin getHandleRaw() {
+        return (net.minecraft.world.entity.monster.piglin.Piglin)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public net.minecraft.world.entity.monster.piglin.Piglin getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (net.minecraft.world.entity.monster.piglin.Piglin) super.getHandle();
     }
 
