--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftThrownPotion.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftThrownPotion.java
@@ -61,8 +_,17 @@
         this.getHandle().splash(null);
     }
     // Paper end
+
+    // Folia start - region threading
+    @Override
+    public net.minecraft.world.entity.projectile.ThrownPotion getHandleRaw() {
+        return (net.minecraft.world.entity.projectile.ThrownPotion)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public net.minecraft.world.entity.projectile.ThrownPotion getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (net.minecraft.world.entity.projectile.ThrownPotion) this.entity;
     }
 }
