--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftFirework.java
@@ -37,8 +_,16 @@
         // Paper end - Expose firework item directly
     }
 
+    // Folia start - region threading
+    @Override
+    public FireworkRocketEntity getHandleRaw() {
+        return (FireworkRocketEntity)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public FireworkRocketEntity getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (FireworkRocketEntity) this.entity;
     }
 
