--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftCreaking.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftCreaking.java
@@ -9,8 +_,16 @@
         super(server, entity);
     }
 
+    // Folia start - region threading
+    @Override
+    public Creaking getHandleRaw() {
+        return (Creaking)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public Creaking getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (Creaking) this.entity;
     }
 
