--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHanging.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHanging.java
@@ -57,8 +_,16 @@
         return CraftBlock.notchToBlockFace(direction);
     }
 
+    // Folia start - region threading
+    @Override
+    public HangingEntity getHandleRaw() {
+        return (HangingEntity)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public HangingEntity getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (HangingEntity) this.entity;
     }
 
