--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderCrystal.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEnderCrystal.java
@@ -39,8 +_,16 @@
         }
     }
 
+    // Folia start - region threading
+    @Override
+    public EndCrystal getHandleRaw() {
+        return (EndCrystal)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public EndCrystal getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (EndCrystal) this.entity;
     }
 
