--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecartTNT.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftMinecartTNT.java
@@ -72,8 +_,16 @@
         this.getHandle().explode(power);
     }
 
+    // Folia start - region threading
+    @Override
+    public net.minecraft.world.entity.vehicle.MinecartTNT getHandleRaw() {
+        return (net.minecraft.world.entity.vehicle.MinecartTNT)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public MinecartTNT getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (MinecartTNT) super.getHandle();
     }
 
