--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftZoglin.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftZoglin.java
@@ -19,8 +_,16 @@
         this.getHandle().setBaby(flag);
     }
 
+    // Folia start - region threading
+    @Override
+    public net.minecraft.world.entity.monster.Zoglin getHandleRaw() {
+        return (net.minecraft.world.entity.monster.Zoglin)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public net.minecraft.world.entity.monster.Zoglin getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (net.minecraft.world.entity.monster.Zoglin) this.entity;
     }
 
