--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPainting.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPainting.java
@@ -50,8 +_,16 @@
         return false;
     }
 
+    // Folia start - region threading
+    @Override
+    public net.minecraft.world.entity.decoration.Painting getHandleRaw() {
+        return (net.minecraft.world.entity.decoration.Painting)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public net.minecraft.world.entity.decoration.Painting getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (net.minecraft.world.entity.decoration.Painting) this.entity;
     }
 
