--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftShulkerBullet.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftShulkerBullet.java
@@ -69,8 +_,16 @@
         return "CraftShulkerBullet";
     }
 
+    // Folia start - region threading
+    @Override
+    public net.minecraft.world.entity.projectile.ShulkerBullet getHandleRaw() {
+        return (net.minecraft.world.entity.projectile.ShulkerBullet)this.entity;
+    }
+    // Folia end - region threading
+
     @Override
     public net.minecraft.world.entity.projectile.ShulkerBullet getHandle() {
+        ca.spottedleaf.moonrise.common.util.TickThread.ensureTickThread(this.entity, "Accessing entity state off owning region's thread"); // Folia - region threading
         return (net.minecraft.world.entity.projectile.ShulkerBullet) this.entity;
     }
 }
