--- a/net/minecraft/server/players/OldUsersConverter.java
+++ b/net/minecraft/server/players/OldUsersConverter.java
@@ -469,7 +_,7 @@
     static Date parseDate(String input, Date defaultValue) {
         Date date;
         try {
-            date = BanListEntry.DATE_FORMAT.parse(input);
+            date = BanListEntry.DATE_FORMAT.get().parse(input); // Folia - region threading - SDF is not thread-safe
         } catch (ParseException var4) {
             date = defaultValue;
         }
