--- a/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
+++ b/net/minecraft/world/entity/projectile/AbstractHurtingProjectile.java
@@ -80,6 +_,11 @@
             this.setPos(location);
             this.applyEffectsFromBlocks();
             super.tick();
+            // Folia start - region threading - make sure entities do not move into regions they do not own
+            if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor((net.minecraft.server.level.ServerLevel)this.level(), this.position(), this.getDeltaMovement(), 1)) {
+                return;
+            }
+            // Folia end - region threading - make sure entities do not move into regions they do not own
             if (this.shouldBurn()) {
                 this.igniteForSeconds(1.0F);
             }
