--- a/net/minecraft/world/entity/projectile/SmallFireball.java
+++ b/net/minecraft/world/entity/projectile/SmallFireball.java
@@ -24,7 +_,7 @@
     public SmallFireball(Level level, LivingEntity owner, Vec3 movement) {
         super(EntityType.SMALL_FIREBALL, owner, movement, level);
         // CraftBukkit start
-        if (this.getOwner() != null && this.getOwner() instanceof Mob) {
+        if (owner != null && this.getOwner() != null && this.getOwner() instanceof Mob) { // Folia - region threading
             this.isIncendiary = (level instanceof ServerLevel serverLevel) && serverLevel.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING);
         }
         // CraftBukkit end
