--- a/net/minecraft/world/level/block/Block.java
+++ b/net/minecraft/world/level/block/Block.java
@@ -362,8 +_,8 @@
             ItemEntity itemEntity = itemEntitySupplier.get();
             itemEntity.setDefaultPickUpDelay();
             // CraftBukkit start
-            if (level.captureDrops != null) {
-                level.captureDrops.add(itemEntity);
+            if (level.getCurrentWorldData().captureDrops != null) { // Folia - region threading
+                level.getCurrentWorldData().captureDrops.add(itemEntity); // Folia - region threading
             } else {
                 level.addFreshEntity(itemEntity);
             }
