--- a/net/minecraft/world/entity/monster/ZombieVillager.java
+++ b/net/minecraft/world/entity/monster/ZombieVillager.java
@@ -69,7 +_,7 @@
     @Nullable
     private MerchantOffers tradeOffers;
     private int villagerXp;
-    private int lastTick = net.minecraft.server.MinecraftServer.currentTick; // CraftBukkit - add field
+    //private int lastTick = net.minecraft.server.MinecraftServer.currentTick; // CraftBukkit - add field // Folia - region threading - restore original timers
 
     public ZombieVillager(EntityType<? extends ZombieVillager> entityType, Level level) {
         super(entityType, level);
@@ -149,7 +_,7 @@
         }
 
         super.tick();
-        this.lastTick = net.minecraft.server.MinecraftServer.currentTick; // CraftBukkit
+        //this.lastTick = net.minecraft.server.MinecraftServer.currentTick; // CraftBukkit // Folia - region threading - restore original timers
     }
 
     @Override
