--- a/src/main/java/org/bukkit/craftbukkit/util/DelegatedGeneratorAccess.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/DelegatedGeneratorAccess.java
@@ -66,6 +_,13 @@
         this.handle = worldAccess;
     }
 
+    // Folia start - region threading
+    @Override
+    public net.minecraft.world.level.StructureManager structureManager() {
+        return this.handle.structureManager();
+    }
+    // Folia end - region threading
+
     public WorldGenLevel getHandle() {
         return this.handle;
     }
@@ -812,4 +_,3 @@
     }
     // Paper end
 }
-
