--- a/net/minecraft/world/entity/raid/Raider.java
+++ b/net/minecraft/world/entity/raid/Raider.java
@@ -86,7 +_,7 @@
             Raid currentRaid = this.getCurrentRaid();
             if (this.canJoinRaid()) {
                 if (currentRaid == null) {
-                    if (this.level().getGameTime() % 20L == 0L) {
+                    if (this.level().getRedstoneGameTime() % 20L == 0L) { // Folia - region threading
                         Raid raidAt = ((ServerLevel)this.level()).getRaidAt(this.blockPosition());
                         if (raidAt != null && Raids.canJoinRaid(this, raidAt)) {
                             raidAt.joinRaid(raidAt.getGroupsSpawned(), this, null, true);
